components/compositing/compositor.rs

```diff
@@ -855,11 +847,22 @@ impl IOCompositor {
                 let mut txn = Transaction::new();
                 for update in updates {
                     match update {
-                        ImageUpdate::AddImage(key, desc, data) => {
+                        ImageUpdate::AddImage(key, desc, data, epoch) => {
+                            if let Some(epoch) = epoch {
+                                assert!(self.image_epochs.insert(key, epoch).is_none());
+                                self.maybe_send_awaiting_display_list();
+                            }
                             txn.add_image(key, desc, data.into(), None)
                         },
-                        ImageUpdate::DeleteImage(key) => txn.delete_image(key),
-                        ImageUpdate::UpdateImage(key, desc, data) => {
+                        ImageUpdate::DeleteImage(key) => {
+                            txn.delete_image(key);
+                            self.image_epochs.remove(&key);
```

Comment:
At which point in the rendering loop are images deleted? 

In particular, can we be sure that if an image for the current epoch is deleted, it will never be contained in the display list for the epoch?

 If the above is not an invariant, then this looks like a kind of livelock, at least for a single epoch, because:

- if an image is deleted here, 
- and we are either waiting on images to update, or on the display list to be received, 
- and the display list where to contain that image key, 
- then the epoch would not be able to process the display list. 

This could be solved by adding a kind of tombstones for the epochs of deleted images, which would require clearing image epochs at the end of an epoch(which could make the logic simpler as well).

Reply:
Hm, I am not sure that's true. At least for webgpu context is not, here we send delete command to painter after context is dropped (gced): https://github.com/servo/servo/blob/7eedeca715eadd4f7c617779b0487ddbdb588090/components/script/dom/webgpu/gpucanvascontext.rs#L404

But I guess that's okay, because at that point there is no canvas anymore.